Add paneling for manually creating a PipelineRun

ncskier opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->
This PR addresses issue https://github.com/tektoncd/dashboard/issues/108 to add the web paneling for manually creating a PipelineRun. There is a new "Create PipelineRun" button at the top of the PipelineRuns page, and each Pipeline's individual page. This button presents a modal form with validation from which the user can create their PipelineRun.

This PR depends on my previous PR https://github.com/tektoncd/dashboard/pull/201 with the PipelinesDropdown container, so please DO NOT MERGE it until that first PR goes in.

Also, @AlanGreene I'm sorry for the large PR, but I wasn't sure how to break down my code further. Please let me know if this PR is too large and if so, how I could go about breaking it down.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

ncskier commented on src/containers/App/App.test.js:
```diff
@@ -18,6 +18,11 @@ import configureStore from 'redux-mock-store';
 import { App } from './App';
+import * as API from '../../api';
+
+beforeEach(() => {
```
Because we get this warning in the tests without it: `console.warn node_modules/fetch-mock/src/lib/fetch-handler.js:73
    Unmatched get to http://localhost/v1/namespaces/undefined/pipelines/` All of the App.js tests use the `getPipelines` function, so I wanted to apply the mock function to all of the tests.

If you'd prefer I apply the mock function individually in each test function, I can make that change.

ncskier commented on src/actions/pipelineRuns.js:
```diff
@@ -50,3 +50,22 @@ export function fetchPipelineRuns({ pipelineName, namespace } = {}) {
+    try {
+      const requestedNamespace = namespace || getSelectedNamespace(getState());
+      headers = await createPipelineRun(payload, requestedNamespace);
+      const pipelineRuns = await getPipelineRuns(
```
I think it returns the PipelineRun and an error: https://godoc.org/github.com/tektoncd/pipeline/pkg/client/clientset/versioned/typed/pipeline/v1alpha1#PipelineRunInterface

However, our dashboard's REST endpoint does not return the PipelineRun

steveodonovan replied:
Should be able to remove this line now. I'd probably go for doing this in the component instead of in an action but I'll leave that up to you, it would be easier and alot less code. If you do keep it as an action change the action type, this isn't fetching so `PIPELINE_RUNS_FETCH_REQUEST` doesn't seem right.

steveodonovan commented:
/ok-to-test

AlanGreene commented on src/actions/pipelineRuns.js:
```diff
@@ -11,7 +11,7 @@ See the License for the specific language governing permissions and
 */
 
-import { getPipelineRun, getPipelineRuns } from '../api';
+import { getPipelineRun, getPipelineRuns, createPipelineRun } from '../api';
```
also `createPipelineRun as createPipelineRunAPI` or similar so we can use `createPipelineRun` as the name of the action creator

re-posting here from our Slack convo so it doesn't get lost

ncskier replied:
Thanks for the feedback. I made the change.

ncskier commented on src/components/TooltipDropdown/TooltipDropdown.js:
```diff
@@ -43,13 +49,15 @@ const TooltipDropdown = ({ items, loading, ...dropdownProps }) => {
   items: [],
-  loading: true
+  loading: false,
+  emptyText: 'No items found'
```
With the addition of the emptyText prop, it made more sense to me that the default state would be an empty dropdown than a loading one.

ncskier commented:
/retest

steveodonovan commented:
Couple of things after trying it out locally, it looks like all pipeline runs are being given names `tekton-pipeline-run-<some uuid>` A combination of the pipeline name and timestamp would be much better here. Well named resources will be important.

Also I see the button to create the pipeline run is at the top of the list. I understand thats definitely better in the case where we have alot of pipeline runs in the list but it does look strange in the UI. Also the advantage of option 2 described in the ticket (having pipeline pre-populated) is gone now given the pipeline runs view. 

![Screenshot 2019-06-10 at 10 35 50](https://user-images.githubusercontent.com/18128367/59187069-6c3a4e00-8b6c-11e9-8c89-589de2bcbf8f.png)

I'm also seeing some strange behaviour where the modal is still around after the its been closed, this might just be a browser inspection thing but would still like @AlanGreene To take a look at the screen shot above.

AlanGreene commented:
Carbon does seem to keep the elements around even after the dialog is closed, reproduced in a codesandbox with standalone Carbon components.

AlanGreene commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -0,0 +1,527 @@
+  'Must consist of only lower case alphanumeric characters, -, and . with at most 253 characters';
+const noSpacesRegExp = /^\S*$/;
+const noSpacesInvalidText = 'Must contain no spaces';
+const urlRegExp = /^(http)s?(:\/\/).*$/;
```
no need for the capturing groups

```diff
-const urlRegExp = /^(http)s?(:\/\/).*$/;
+const urlRegExp = /^https?:\/\/.*$/;
```

This will allow a 'URL' such as `http://` or `https://`, but changing the kleene star to a plus would probably give _slightly_ more useful validation by enforcing at least _some_ content after the //

`const urlRegExp = /^https?:\/\/.+/;`

steveodonovan commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -0,0 +1,527 @@
+
+const mapStateToProps = state => {
+  return {
+    errorMessage: getPipelineRunsErrorMessage(state),
```
These aren't used. Id say make call directly from the component and handle the error here

steveodonovan commented on src/containers/CreatePipelineRun/CreatePipelineRun.scss:
```diff
@@ -0,0 +1,43 @@
+limitations under the License.
+*/
+
+// @import '~carbon-components/scss/globals/scss/styles';
```
remove comments

steveodonovan commented on src/containers/CreatePipelineRun/CreatePipelineRun.test.js:
```diff
@@ -0,0 +1,404 @@
+      <CreatePipelineRun open />
+    </Provider>
+  );
+  document.getElementById(
```
you can get rid of all of these and their `const scrollIntoViewMock = jest.fn();`

ncskier replied:
Sorry about that. I forgot to remove them.

ncskier commented on src/containers/CreatePipelineRun/CreatePipelineRun.scss:
```diff
@@ -0,0 +1,43 @@
+// @import '../../scss/vars';
+
+.create-pipelinerun {
+  .bx--modal-header,
```
In what file do we store global styles? Should I move it to `App.scss`?

AlanGreene replied:
App.scss is fine for now 👍 

I've been doing some work locally to try clean things up a bit but won't be ready for a while, ties into some Carbon update stuff that's proving a pain at the moment.

ncskier commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -0,0 +1,527 @@
+    const { required, regexp } = formValidation[key];
+    const invalid =
+      (required && value === '') ||
+      (!!regexp && !regexp.test(value) && !(!required && value === ''));
```
This boolean expression isn't the same. It doesn't achieve the logic I want with the required values. I want something to be invalid when the field is `required`, and the `value` is empty `''`
EDIT
My mistake, you're probably only referring to the second line, but I thought you meant the entire expression 😄 
I'll make it:
```
const invalid =
      (required && value === '') ||
      (regexp && !regexp.test(value) && (required || value !== '');
```

AlanGreene replied:
yeah sorry should have made that clearer, it was just the second line
your update looks good

ncskier commented on src/containers/CreatePipelineRun/CreatePipelineRun.test.js:
```diff
@@ -0,0 +1,404 @@
+      }
+    })
+  );
+  fireEvent.click(getByText(/submit/i));
```
I updated the test to expect the function to be called:
```
const createPipelineRun = jest
    .spyOn(API, 'createPipelineRun')
    .mockImplementation(() =>
      Promise.resolve({
        get: () => {
          return '/v1/namespaces/default/pipelineruns//tekton-pipeline-run';
        }
      })
    );
  fireEvent.click(getByText(/submit/i));
  await wait(() => expect(createPipelineRun).toHaveBeenCalledTimes(1));
```
However, I do not know how to assert that the parameters are correct. When I put `expect().toBe();` statements in the `mockImplementation`, the test doesn't fail when the `expect` fails. Do you know how I can assert the correct parameters?

steveodonovan replied:
`expect(createPipelineRun).toHaveBeenCalledWith(<expected params>)`

ncskier replied:
Ah, thank you! 😄

ncskier commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -0,0 +1,527 @@
+                    />
+                  );
+                }
+                return null;
```
ESLint complains if I don't have it :/
`Expected to return a value at the end of arrow function.eslint(consistent-return)`

AlanGreene replied:
instead of this:

```js
{(() => {
  if (this.state.createdPipelineRun) {
    return (
      <InlineNotification
        kind="success"
        title="Successfully created a PipelineRun"
        subtitle={
          <Link to={this.state.createdPipelineRunURL}>
            {this.state.createdPipelineRunName}
          </Link>
        }
        hideCloseButton={false}
      />
    );
  }
  return null;
})()}
```

you should be able to do:

```js
{this.state.createdPipelineRun && (
  <InlineNotification
    kind="success"
    title="Successfully created a PipelineRun"
    subtitle={
      <Link to={this.state.createdPipelineRunURL}>
        {this.state.createdPipelineRunName}
      </Link>
    }
    hideCloseButton={false}
  />
)}
```

ncskier replied:
Awesome, thank you! 😄

ncskier commented:
@steveodonovan @AlanGreene thank you so much for your thorough review! I think I have addressed all of your concerns. The largest change I made was removing the `createPipelineRun` Action, and instead using the API directly (I have added new tests for this). Also, I talked with Steve on Slack about changing  the PipelineRun name to use a timestamp uid; I will deliver this change in a separate PR.

Please let me know if there are any more modifications I can make to improve the code 🙂

AlanGreene commented on src/containers/PipelineRuns/PipelineRuns.js:
```diff
@@ -73,6 +93,53 @@ export /* istanbul ignore next */ class PipelineRuns extends Component {
 
     return (
       <>
+        {(() => {
```
similar to the others in CreatePipelineRun, `return null` not required

ncskier replied:
Thanks. I missed that one 👍

AlanGreene commented on src/containers/PipelineRuns/PipelineRuns.js:
```diff
@@ -73,6 +93,53 @@ export /* istanbul ignore next */ class PipelineRuns extends Component {
+        </Button>
+        {(() => {
+          let namespace;
+          if (pipelineName) {
```
move this above the return, it'll clean up the JSX so you can just have the `<CreatePipelineRun... />` directly without needing this function and all the associated brackets

AlanGreene commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -0,0 +1,523 @@
+    const pipelineValue = this.getPipeline();
+
+    return (
+      <>
```
no need for a fragment when returning a single element

ncskier commented on src/containers/PipelineRuns/PipelineRuns.js:
```diff
@@ -21,8 +21,10 @@ import {
   StructuredListSkeleton,
-  StructuredListWrapper
+  StructuredListWrapper,
+  Button
```
Yeah, automation for this is pretty necessary.

AlanGreene replied:
I'll see if this is something me might be able to do with prettier

ncskier replied:
Thank you! 😄

ncskier commented on src/containers/PipelineRuns/PipelineRuns.js:
```diff
@@ -73,6 +93,53 @@ export /* istanbul ignore next */ class PipelineRuns extends Component {
+        </Button>
+        {(() => {
+          let namespace;
+          if (pipelineName) {
```
No, there's no `pipelineName` in the case when we're on the main `PipelineRuns` page, because there isn't a specific pipeline selected.

ncskier commented:
/retest

ncskier commented:
@AlanGreene @steveodonovan Thank you for the additional feedback. I have addressed your concerns, and I'll follow up with Steve on Slack about the PipelineRun resource names.

Let me know if there's anything else I can change 🙂

AlanGreene commented on src/containers/CreatePipelineRun/CreatePipelineRun.js:
```diff
@@ -0,0 +1,521 @@
+import './CreatePipelineRun.scss';
+import { ALL_NAMESPACES } from '../../constants';
+
+const resourceNameRegExp = /^[-.a-z1-9]{1,253}$/;
```
this excludes '0', is that intentional?

ncskier replied:
It is not 😅
